File: price_algorithm.js
const PROBES = [375, 583, 585, 750, 850, 875, 916, 999];
const HIDDEN_PROBES = [900];

// Source: Zoloto.xlsx, main rate 870.
// Keep the workbook's "to minus from" spread per brand/probe instead of
// freezing a "to" value that becomes invalid when the main rate changes.
// Tillachi Bolla uses the same public ranges as Diamant.
const BRAND_MAX_ADDITIONS = {
  skupka: {
    375: 70000,
    583: 330000,
    585: 180000,
    750: 380000,
    850: 75000,
    875: 75000,
    900: 155000,
    916: 90000,
    999: 105000,
  },
  diamant: {
    375: 70000,
    583: 340000,
    585: 180000,
    750: 380000,
    850: 75000,
    875: 85000,
    900: 155000,
    916: 60000,
    999: 120000,
  },
  tillachi: {
    375: 70000,
    583: 340000,
    585: 180000,
    750: 380000,
    850: 75000,
    875: 85000,
    900: 155000,
    916: 60000,
    999: 120000,
  },
  goldexpert: {
    375: 70000,
    583: 350000,
    585: 180000,
    750: 380000,
    850: 70000,
    875: 90000,
    900: 155000,
    916: 80000,
    999: 95000,
  },
};

function excelRound(value, digits = 0) {
  const factor = 10 ** digits;
  return Math.sign(value) * Math.round(Math.abs(value) * factor) / factor;
}

function excelCeiling(value, significance) {
  return Math.ceil(value / significance) * significance;
}

function roundupTo10000(value) {
  return excelCeiling(value, 10000);
}

function roundedBasePrice(probe, mainRate) {
  return excelCeiling((probe / 583 / 10) * mainRate, 0.5) * 10000;
}

function minPrice(probe, mainRate) {
  if (probe === 585) {
    return Math.round(mainRate * 1000);
  }
  return roundedBasePrice(probe, mainRate);
}

function maxPrice(probe, startPrice, brand) {
  const additions = BRAND_MAX_ADDITIONS[brand] || BRAND_MAX_ADDITIONS.diamant;
  const addition = additions[probe] || BRAND_MAX_ADDITIONS.diamant[probe] || 0;
  return startPrice + addition;
}

function calculatePrices(mainRate, brand = "diamant", includeHidden = false) {
  const rate = Number(mainRate);
  if (!Number.isFinite(rate) || rate <= 0) {
    throw new Error("mainRate must be a positive number");
  }

  const result = {};
  const probes = includeHidden ? [...PROBES, ...HIDDEN_PROBES] : PROBES;
  const normalizedBrand = String(brand || "diamant").toLowerCase();
  for (const probe of probes) {
    const start = minPrice(probe, rate);
    result[String(probe)] = [start, maxPrice(probe, start, normalizedBrand)];
  }
  return result;
}

function generatePriceRange(mainRate, brand = "diamant") {
  return calculatePrices(mainRate, brand);
}

function formatPrice(value) {
  return Math.round(value).toString().replace(/\B(?=(\d{3})+(?!\d))/g, ".");
}

module.exports = {
  calculatePrices,
  generatePriceRange,
  formatPrice,
};

if (require.main === module) {
  const mainRate = Number(process.argv[2] || 1200);
  console.log(JSON.stringify(generatePriceRange(mainRate), null, 2));
}

File: price_algorithm.py
import math


PROBAS = (375, 583, 585, 750, 850, 875, 916, 999)
# Existing bot modules import PROBES. Keep this alias for compatibility.
PROBES = PROBAS

BRAND_EXTRA = {
    "skupka": 70_000,
    "diamant": 75_000,
    "tillachi_bolla": 75_000,
    "goldexpert": 80_000,
}

# The deployed Tillachi endpoint still uses this legacy identifier.
BRAND_ALIASES = {"tillachi": "tillachi_bolla"}


def calculate_start_price(proba: int, main_rate: int) -> int:
    if proba == 585:
        return main_rate * 1000

    raw = (proba / 583 / 10) * main_rate
    return int(math.ceil(raw * 2) / 2 * 10000)


def calculate_prices(main_rate: int, brand: str) -> dict:
    brand = BRAND_ALIASES.get(str(brand).lower(), str(brand).lower())
    if brand not in BRAND_EXTRA:
        raise ValueError(f"Неизвестная компания: {brand}")

    main_rate = int(main_rate)
    if main_rate <= 0:
        raise ValueError("main_rate must be positive")

    extra = BRAND_EXTRA[brand]
    prices = {}

    for proba in PROBAS:
        from_price = calculate_start_price(proba, main_rate)

        if proba == 583:
            to_price = from_price + extra + 260_000
        elif proba == 585:
            to_price = from_price + 180_000
        elif proba == 750:
            to_price = 1_500_000
        else:
            to_price = from_price + extra

        prices[proba] = {"from": int(from_price), "to": int(to_price)}

    return prices


def generate_price_range(main_rate: int, brand: str = "diamant") -> dict:
    return calculate_prices(main_rate, brand)


def format_price(value: int) -> str:
    return f"{int(value):,}".replace(",", ".")


def price_rows(main_rate: int, brand: str = "diamant") -> list:
    ranges = calculate_prices(main_rate, brand)
    return [
        (
            str(proba),
            f"{format_price(price_range['from'])}-{format_price(price_range['to'])} сум/гр",
        )
        for proba, price_range in ranges.items()
    ]
